File: LeetCode/450. Delete Node in a BST/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        if not root:
            return root
        if root.val == key:
            left = self.findMax(root.left)
            if left:
                left.right = root.right
                if left != root.left:
                    left.left = root.left
                root = left
            else:
                root = root.right
        elif root.val > key:
            root.left = self.deleteNode(root.left, key)
        else:
            root.right = self.deleteNode(root.right,key)
        return root
    
    def findMax(self, root):
        if root and root.right:
            if root.right:
                right = self.findMax(root.right)
                if right == root.right:
                    root.right = right.left
                return right

        return root






# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        def dfs(cur):
            if not cur:
                return None
            if cur.val == key:                
                tmp = cur
                pre, cur = None, cur.left
                if not cur:
                    return tmp.right
                if not cur.right:
                    cur.right = tmp.right
                    return cur
                while cur.right:
                    pre = cur
                    cur = cur.right
                pre.right = cur.left
                cur.left = tmp.left
                cur.right = tmp.right
                return cur
            if key < cur.val:
                cur.left = dfs(cur.left)
            else:
                cur.right = dfs(cur.right)
            return cur
        return dfs(root)

File: LeetCode/450. Delete Node in a BST/solution.cpp
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) 
    {
        if (root == nullptr)    return nullptr;
        if (key > root->val)    root->right = deleteNode(root->right, key);     // 去右子树删除
        else if (key < root->val)    root->left = deleteNode(root->left, key);  // 去左子树删除
        else    // 当前节点就是要删除的节点
        {
            if (! root->left)   return root->right; // 情况1，欲删除节点无左子
            if (! root->right)  return root->left;  // 情况2，欲删除节点无右子
            TreeNode* node = root->right;           // 情况3，欲删除节点左右子都有 
            while (node->left)          // 寻找欲删除节点右子树的最左节点
                node = node->left;
            node->left = root->left;    // 将欲删除节点的左子树成为其右子树的最左节点的左子树
            root = root->right;         // 欲删除节点的右子顶替其位置，节点被删除
        }
        return root;    
    }
};


// 作者：Terry2020
// 链接：https://leetcode.cn/problems/delete-node-in-a-bst/solution/miao-dong-jiu-wan-shi-liao-by-terry2020-tc0o/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。